### DISCUSSION TITLE
Attest the application code before enclave building

### DISCUSSION THREAD

**Discussion by ROF13ThFloor** [gramineproject/gramine#1853] [Q&A]
Hi all, I did not find the related thread, so I'm sorry if it's a repetitive question

I have a Node.js application which is deployed on the application. After deployment, I create the enclave in the cloud (Azure) and then before the execution Gramine use attestation to attest the application and codes. 

My question is, Does attestation verify the application code before enclave creation? 

In my deployment process, I copied the application to the target VM and then creat the enclave inside the cloud. But it seems that the attestation can verify the created enclave and if the code changes before building the enclave, then the building enclave and attestation works! 

So my question is, is there any way to protect from code before enclave building? Is it possible to build an enclave before sending the application to the cloud? 

I really appreciate any answer to my question 

Thanks

**Comment by dimakuv** [comment]
Sorry, could you reformulate your question in a more precise manner? In particular, with code snippets of what you're doing.

For example, what do you mean by "enclave building"? What do you mean by "I use attestation"?

In general, the correct way is to:
1. Prepare a correct manifest file (`.manifest.template`) for your app, that contains all code/binaries/configs as `sgx.trusted_files`.
2. Build the Gramine files (`.manifest.sgx`, `.sig`) on your local machine. Achieved using Gramine tools like `gramine-manifest` and `gramine-sgx-sign`.
3. Package all the files in some way (typically using a Docker image).
4. Send the packaged files to the remote machine/VM.
5. Start the packaged files via `gramine-sgx ...` on that remote machine. As part of that process, remote attestation must be performed (you must incorporate RA-TLS or smth like that in step 1).

After these steps, when you receive the attestation report on your local machine and verify the MRENCLAVE, you can be sure that the code of the app was not modified.

**Reply by ROF13ThFloor** [reply]
Hi, Thank you for your answer. Yes sure I will explain it more 

I have a Node.js application with a set of Directories and Files.

1. First I deployed the application in Azure by using the Ansible deployment tool (using `ssh` connection ) and also I will send Gramine configuration including the `manifest.toml` file 
2. Then In the related directory I will build the enclave by running this command: `make SGX=1 SGX_SIGNER_KEY=PathToEnclaveKey DEBUG=1` 
3. Then I will start the application by running this command: `gramine-sgx node main.js `

During the execution application will communicate with the attestation verifier. 

As far as I understand based on your suggestion, I need to create `.sig` files on a local machine and then send it to the cloud.

**Reply by dimakuv** [reply]
> As far as I understand based on your suggestion, I need to create .sig files on a local machine and then send it to the cloud.

Yes.

Don't forget that you also need to perform remote attestation in a proper way. Check https://gramine.readthedocs.io/en/stable/attestation.html

**Reply by ROF13ThFloor** [reply]
Thanks for your answer. 

My last question is about the security of this type of deployment.  As far as I understand we do not pass the `.sig` file to the attestation verifier.  So How can I make sure that the cloud can not generate it again with modified application ? 


In the provided link, there's a mention of expected measurements, but I'm uncertain about the method for storing these measurements for the application code. Would additional configuration in the attestation verifier be necessary to verify pre-generated .sig files?
Thanks

**Reply by dimakuv** [reply]
> So How can I make sure that the cloud can not generate it again with modified application ?

If the application is modified, then the MRENCLAVE measurement will be different.

Also, the cloud admin doesn't have access to your private enclave-signing key, so MRSIGNER measurement will also be different.

With these two measurements being different from the expected values, remote attestation will fail on your attestation-verifier side.

> Would additional configuration in the attestation verifier be necessary to verify pre-generated .sig files?

You shouldn't care about `.sig` files. They are not used during attestation. They are used purely during startup of the SGX enclave, and the sole reason these files exist is the SGX sealing capability.

I would highly recommend to read more literature about SGX remote attestation.
